### PR TITLE
Add json and inferred format

### PR DESCRIPTION
**Pull request by sfc-gh-jcarroll**
- Add support for `input_format='json'` and `input_format='jsonl'` to read() method
- Try to infer input_format from Path.suffix if not specified

### CODE REVIEW

**Comment by sfc-gh-jcarroll** [comment]
@sfc-gh-zblackwood @sfc-gh-tszerszen @vdonato any chance someone could take a look at this soon? Has been sitting for a bit now

**Review by sfc-gh-zblackwood — COMMENTED** [review]
My only concern with this approach is that there are many cases (perhaps the majority?) where I actually want to just return a raw object (dict, list, etc.) from a json file, rather than a DataFrame. I don't know a great way to solve this, but perhaps an argument for `as_dataframe` or something like that. But, that doesn't work well for non-json files.

Or, perhaps there's a special input_format called `raw_json` that is specifically used to mean "don't return a dataframe, just a normal object from the JSON"

**Comment by sfc-gh-jcarroll** [comment]
That makes sense. We talked about adding a `format=` argument to specify the output format - I think that's the best way to solve this? relevant for dict/list as well as if you want a pyarrow.Table for tabular (or later polars or whatever else)

Would be open to try that out here (although may be a bit before I have time to hack it together). Should the default read method change output format depending on the input format?

**Comment by sfc-gh-zblackwood** [comment]
@sfc-gh-jcarroll Yes, that seems like a good way to handle that -- I forgot we had talked about that argument.

I think the default read method already changes output format sometimes, e.g. .txt returns a string, right?

**Comment by sfc-gh-jcarroll** [comment]
Ah, yep, you're right! So should json == dict by default and jsonl == list by default?

Nice thing is we can still pass those to st.dataframe I think :)

**Comment by sfc-gh-jcarroll** [comment]
Also - any idea on the best way to implement it?

File: st_files_connection/connection.py [hunk 1]
```diff
@@ -151,18 +161,38 @@ def _read_csv(path: str | Path, **kwargs) -> pd.DataFrame:
...
+            if "connection_name" in kwargs:
+                kwargs.pop("connection_name")
+
+            with self.open(path, "rt") as f:
+                return pd.read_json(f, **kwargs)
+
+        # Try to infer input_format from file extension if missing
+        if input_format is None:
+            # You can construct a Path from a Path so this should work regardless
+            input_format = Path(path).suffix.replace('.', '', 1)
```
Comment (sfc-gh-zblackwood) [inline st_files_connection/connection.py:183]:
```suggestion
            input_format = Path(path).suffix.replace('.', '', 1)
            if input_format == 'txt':
                input_format = 'text'
```

**Comment by sfc-gh-zblackwood** [comment]
Adding type hints for json is hard -- last I checked, python's built-in json.load just has a hint that it returns `Any`. I would guess that jsonl files tend to be more tabular datasets, but that's just a guess. If that's true, than just returning a dataframe might be reasonable for jsonl. I would be a bit hesitant to be too confident about typing what json returns generically, but if you pass `pandas` or `dict` or `list` or something, then you should expect to get those. I guess I would imagine something like this:

```python
    @overload
    def read(
        self,
        path: str | Path,
        input_format: Literal["json"],
        ttl: Optional[Union[float, int, timedelta]] = None,
        **kwargs,
    ) -> Any:
        pass
        
    @overload
    def read(
        self,
        path: str | Path,
        input_format: Literal["json"],
        format: Literal["pandas"],
        ttl: Optional[Union[float, int, timedelta]] = None,
        **kwargs,
    ) -> pd.DataFrame:
        pass
        
    @overload
    def read(
        self,
        path: str | Path,
        input_format: Literal["json"],
        format: Literal["dict"],
        ttl: Optional[Union[float, int, timedelta]] = None,
        **kwargs,
    ) -> dict:
        pass
        
    ...
    # And then checking for various combinations of arguments when actually creating the read method
```

**Review by vdonato — COMMENTED** [review]
Hm, I'm not entirely sure what I think about `json` handling.

A `format` arg would solve this but would likely be a ton of work to implement generally. I think that for `json` files specifically, most of the time you'll want a normal python dict / list of dicts rather than a dataframe, and since pandas should already have methods for converting from dict -> df, there's a reasonable argument for just returning a dict for `json` inputs

**Comment by sfc-gh-jcarroll** [comment]
Updated as suggested, LMK if this looks better!

**Review by sfc-gh-zblackwood — APPROVED** [review]
Looks great to me

**Pull request merged**